circup/backends.py

```diff
@@ -0,0 +1,660 @@
+            click.echo("No module name(s) provided.")
+        elif name in mod_names:
+            # Grab device modules to check if module already installed
+            if name in device_modules:
+                click.echo("'{}' is already installed.".format(name))
+                return
+
+            library_path = os.path.join(device_path, self.LIB_DIR_PATH)
+
+            # Create the library directory first.
+            self._create_library_directory(device_path, library_path)
+
+            metadata = mod_names[name]
+            bundle = metadata["bundle"]
+            if pyext:
+                # Use Python source for module.
+                self._install_module_py(metadata)
+            else:
+                # Use pre-compiled mpy modules.
+                self._install_module_mpy(bundle, metadata)
+            click.echo("Installed '{}'.".format(name))
+        else:
+            click.echo("Unknown module named, '{}'.".format(name))
+
+    # def libraries_from_imports(self, code_py, mod_names):
+    #     """
+    #     To be overridden by subclass
+    #     """
+    #     raise NotImplementedError
+
+    def uninstall(self, device_path, module_path):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def update(self, module):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def get_file_path(self, filename):
+        """
+        To be overridden by subclass
+        """
+        raise NotImplementedError
+
+    def is_device_present(self):
+        """
+        To be overriden by subclass
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def parse_boot_out_file(boot_out_contents):
+        """
+        Parse the contents of boot_out.txt
+        Returns: circuitpython version and board id
+        """
+        lines = boot_out_contents.split("\n")
+        version_line = lines[0]
+        circuit_python = version_line.split(";")[0].split(" ")[-3]
+        board_line = lines[1]
+        if board_line.startswith("Board ID:"):
+            board_id = board_line[9:].strip()
+        else:
+            board_id = ""
+        return circuit_python, board_id
+
+
+class WebBackend(Backend):
+    """
+    Backend for interacting with a device via Web Workflow
+    """
+
+    def __init__(self, host, password, logger):
+        super().__init__(logger)
+        self.LIB_DIR_PATH = "fs/lib/"
+        self.host = host
+        self.password = password
+        self.device_location = f"http://:{self.password}@{self.host}"
+
+        self.library_path = self.device_location + "/" + self.LIB_DIR_PATH
+
+    def install_file_http(self, source):
+        """
+        Install file to device using web workflow.
+        :param source source file.
+        """
+
+        target = (
+            self.device_location
+            + "/"
+            + self.LIB_DIR_PATH
+            + source.split(os.path.sep)[-1]
```

Comment:
How does this work when the source is nested in a folder? Maybe we need to pass the full path and the target subpath?

Reply:
I think it didn't work correctly as it was written. This bit will be changed to this with some of the changes from @tyeth that I am working on merging in:
```
        mod_name = source.split(os.path.sep)
        mod_name = mod_name[-2] if mod_name[-1] == "" else mod_name[-1]
        target = (
            self.device_location
            + "/"
            + self.LIB_DIR_PATH
            + mod_name
        )
```